src/pudl/transform/ferc1.py

```diff
@@ -960,75 +1124,332 @@ def calculate_values_from_components(
+            )
+
+        metric_name = self.snake_case_metric_name()
+        df = (
+            pd.DataFrame(self.calculate_metric(calculated_df), columns=[metric_name])
+            .assign(
+                **{  # totolerance_ is just for reporting so you can know of off you are
+                    f"tolerance_{metric_name}": self.metric_tolerance,
+                    f"is_error_{metric_name}": lambda x: x[metric_name]
+                    > self.metric_tolerance,
+                }
+            )
+            # make a uniform multi-index w/ group name and group values
+            .assign(group=self.by)
+            .reset_index()
+            .rename(columns={self.by: "group_value"})
+        )
+        # we want to set the index values as the same for all groups, but both the
+        # ungrouped and table_name group require a special exception because we need
+        # to add table_name into the columns
+        if self.by == "ungrouped":
+            df = df.assign(table_name="ungrouped")
+        if self.by == "table_name":
+            # we end up having two columns w/ table_name values in order to keep all the
+            # outputs having the same indexes.
+            df = df.assign(table_name=lambda x: x["group_value"])
+        return df.set_index(["group", "table_name", "group_value"])
+
+
+class ErrorFrequency(GroupMetricCheck):
+    """Check error frequency in XBRL calculations."""
+
+    def metric(self: Self, gb: DataFrameGroupBy) -> pd.Series:
+        """Calculate the frequency with which records are tagged as errors."""
+        try:
+            out = gb[gb.is_not_close].shape[0] / gb.shape[0]
+        except ZeroDivisionError:
+            # Will only occur if all reported values are NaN when calculated values
+            # exist, or vice versa.
+            logger.warning(
+                "Calculated values have no corresponding reported values in this table."
+            )
+            out = np.nan
+        return out
+
+
+class RelativeErrorMagnitude(GroupMetricCheck):
+    """Check relative magnitude of errors in XBRL calculations."""
+
+    def metric(self: Self, gb: DataFrameGroupBy) -> pd.Series:
+        """Calculate the mangnitude of the errors relative to total reported value."""
+        try:
+            # Should we be taking the absolute value of the reported column?
+            return gb[gb.is_not_close].abs_diff.sum() / gb["reported_value"].abs().sum()
+        except ZeroDivisionError:
+            return np.nan
+
+
+class AbsoluteErrorMagnitude(GroupMetricCheck):
+    """Check relative magnitude of errors in XBRL calculations.
+
+    These numbers may vary wildly from table to table so no default values for the
+    expected errors are provided here...
+    """
+
+    def metric(self: Self, gb: DataFrameGroupBy) -> pd.Series:
+        """Calculate the absolute mangnitude of XBRL calculation errors."""
+        return gb.abs_diff.sum()
+
+
+class NullCalculationFrequency(GroupMetricCheck):
```

Comment:
To match the name naming of the next class below, this should probably be called something like `NullCalculatedValueFrequency`

Clarify in the docstring that we're calculating the frequency relative to the number of non-null reported values.

"""Fraction of non-null reported values that have null corresponding calculated values"""